Problems compiling dg with llvm 6.0.1 on OSX

Hi all,

I am trying to compile **dg** with LLVM 6.0.1 but the make part fails with the following error:

```
Undefined symbols for architecture x86_64:
  "llvm::DataLayout::~DataLayout()", referenced from:
      dg::analysis::pta::LLVMPointerSubgraphBuilder::~LLVMPointerSubgraphBuilder() in PointerSubgraph.cpp.o
  "llvm::raw_ostream::write(char const*, unsigned long)", referenced from:
      llvm::raw_ostream::operator<<(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) in PointerSubgraph.cpp.o
      llvm::raw_ostream::operator<<(llvm::StringRef) in PointerSubgraph.cpp.o
      llvm::raw_ostream::operator<<(llvm::StringRef) in Globals.cpp.o
      llvm::raw_ostream::operator<<(llvm::StringRef) in Constants.cpp.o
      llvm::raw_ostream::operator<<(llvm::StringRef) in Instructions.cpp.o
      llvm::raw_ostream::operator<<(llvm::StringRef) in Calls.cpp.o
  "llvm::raw_ostream::operator<<(unsigned long)", referenced from:
      llvm::raw_ostream::operator<<(unsigned int) in Constants.cpp.o
      llvm::raw_ostream::operator<<(unsigned int) in Instructions.cpp.o
  "llvm::ConstantExpr::getAsInstruction()", referenced from:
      dg::analysis::pta::LLVMPointerSubgraphBuilder::getConstantExprPointer(llvm::ConstantExpr const*) in Constants.cpp.o
  "llvm::PassRegistry::getPassRegistry()", referenced from:
      llvm::DominatorTreeWrapperPass::DominatorTreeWrapperPass() in PointerSubgraph.cpp.o
  "llvm::DomTreeNodeBase<llvm::BasicBlock>::end()", referenced from:
      dg::analysis::pta::getBasicBlocksInDominatorOrder(llvm::Function&) in PointerSubgraph.cpp.o
  "llvm::DomTreeNodeBase<llvm::BasicBlock>::begin()", referenced from:
      dg::analysis::pta::getBasicBlocksInDominatorOrder(llvm::Function&) in PointerSubgraph.cpp.o
  "llvm::DominatorTreeBase<llvm::BasicBlock, false>::getRootNode()", referenced from:
      dg::analysis::pta::getBasicBlocksInDominatorOrder(llvm::Function&) in PointerSubgraph.cpp.o
  "llvm::DominatorTreeBase<llvm::BasicBlock, false>::DominatorTreeBase()", referenced from:
      llvm::DominatorTree::DominatorTree() in PointerSubgraph.cpp.o
  "llvm::EnableABIBreakingChecks", referenced from:
      llvm::VerifyEnableABIBreakingChecks in PointerSubgraph.cpp.o
      llvm::VerifyEnableABIBreakingChecks in PointerSubgraphValidator.cpp.o
      llvm::VerifyEnableABIBreakingChecks in Structure.cpp.o
      llvm::VerifyEnableABIBreakingChecks in Globals.cpp.o
      llvm::VerifyEnableABIBreakingChecks in Constants.cpp.o
      llvm::VerifyEnableABIBreakingChecks in Instructions.cpp.o
      llvm::VerifyEnableABIBreakingChecks in Calls.cpp.o
      ...
  "llvm::DominatorTreeWrapperPass::runOnFunction(llvm::Function&)", referenced from:
      dg::analysis::pta::getBasicBlocksInDominatorOrder(llvm::Function&) in PointerSubgraph.cpp.o
  "llvm::DominatorTreeWrapperPass::ID", referenced from:
      llvm::DominatorTreeWrapperPass::DominatorTreeWrapperPass() in PointerSubgraph.cpp.o
  "llvm::llvm_unreachable_internal(char const*, char const*, unsigned int)", referenced from:
      llvm::DataLayout::getTypeSizeInBits(llvm::Type*) const in PointerSubgraph.cpp.o
      llvm::DataLayout::getTypeSizeInBits(llvm::Type*) const in Globals.cpp.o
      llvm::DataLayout::getTypeSizeInBits(llvm::Type*) const in Instructions.cpp.o
  "llvm::initializeDominatorTreeWrapperPassPass(llvm::PassRegistry&)", referenced from:
      llvm::DominatorTreeWrapperPass::DominatorTreeWrapperPass() in PointerSubgraph.cpp.o
  "llvm::Pass::~Pass()", referenced from:
      llvm::FunctionPass::~FunctionPass() in PointerSubgraph.cpp.o
  "llvm::User::getDescriptor()", referenced from:
      llvm::OperandBundleUser<llvm::CallInst, llvm::Use*>::bundle_op_info_begin() in PointerSubgraph.cpp.o
      llvm::OperandBundleUser<llvm::CallInst, llvm::Use*>::bundle_op_info_end() in PointerSubgraph.cpp.o
  "llvm::errs()", referenced from:
      dg::analysis::pta::dump(llvm::BasicBlock const&) in PointerSubgraph.cpp.o
      dg::analysis::pta::dump(llvm::Instruction const&) in PointerSubgraph.cpp.o
      dg::analysis::pta::dump(llvm::Value const&) in PointerSubgraph.cpp.o
      dg::analysis::pta::LLVMPointerSubgraphBuilder::tryGetOperand(llvm::Value const*) in PointerSubgraph.cpp.o
      dg::analysis::pta::LLVMPointerSubgraphBuilder::getOperand(llvm::Value const*) in PointerSubgraph.cpp.o
      dg::analysis::pta::LLVMPointerSubgraphBuilder::buildInstruction(llvm::Instruction const&) in PointerSubgraph.cpp.o
      dg::analysis::pta::LLVMPointerSubgraphBuilder::checkMemSet(llvm::Instruction const*) in PointerSubgraph.cpp.o
      ...
  "llvm::APInt::initSlowCase(unsigned long long, bool)", referenced from:
      llvm::APInt::APInt(unsigned int, unsigned long long, bool) in Constants.cpp.o
      llvm::APInt::APInt(unsigned int, unsigned long long, bool) in Instructions.cpp.o
  "llvm::Value::deleteValue()", referenced from:
      dg::analysis::pta::LLVMPointerSubgraphBuilder::getConstantExprPointer(llvm::ConstantExpr const*) in Constants.cpp.o
  "llvm::AllocaInst::isArrayAllocation() const", referenced from:
      dg::analysis::pta::getAllocatedSize(llvm::AllocaInst const*, llvm::DataLayout const*) in Instructions.cpp.o
  "llvm::BasicBlock::getTerminator() const", referenced from:
      llvm::succ_begin(llvm::BasicBlock const*) in Structure.cpp.o
      llvm::succ_end(llvm::BasicBlock const*) in Structure.cpp.o
  "llvm::DataLayout::getPointerSize(unsigned int) const", referenced from:
      llvm::DataLayout::getPointerSizeInBits(unsigned int) const in PointerSubgraph.cpp.o
      llvm::DataLayout::getPointerSizeInBits(unsigned int) const in Globals.cpp.o
      llvm::DataLayout::getPointerSizeInBits(unsigned int) const in Constants.cpp.o
      llvm::DataLayout::getPointerSizeInBits(unsigned int) const in Instructions.cpp.o
  "llvm::DataLayout::getStructLayout(llvm::StructType*) const", referenced from:
      llvm::DataLayout::getTypeSizeInBits(llvm::Type*) const in PointerSubgraph.cpp.o
      llvm::DataLayout::getTypeSizeInBits(llvm::Type*) const in Globals.cpp.o
      dg::analysis::pta::accumulateEVOffsets(llvm::ExtractValueInst const*, llvm::DataLayout const&) in Instructions.cpp.o
      llvm::DataLayout::getTypeSizeInBits(llvm::Type*) const in Instructions.cpp.o
  "llvm::DataLayout::getABITypeAlignment(llvm::Type*) const", referenced from:
      llvm::DataLayout::getTypeAllocSize(llvm::Type*) const in PointerSubgraph.cpp.o
      llvm::DataLayout::getTypeAllocSize(llvm::Type*) const in Globals.cpp.o
      llvm::DataLayout::getTypeAllocSize(llvm::Type*) const in Instructions.cpp.o
  "llvm::GlobalValue::isDeclaration() const", referenced from:
      dg::analysis::pta::LLVMPointerSubgraphBuilder::buildFunction(llvm::Function const&) in PointerSubgraph.cpp.o
      llvm::GlobalVariable::hasInitializer() const in Globals.cpp.o
  "llvm::CompositeType::indexValid(unsigned int) const", referenced from:
      dg::analysis::pta::accumulateEVOffsets(llvm::ExtractValueInst const*, llvm::DataLayout const&) in Instructions.cpp.o
  "llvm::CompositeType::getTypeAtIndex(unsigned int) const", referenced from:
      dg::analysis::pta::accumulateEVOffsets(llvm::ExtractValueInst const*, llvm::DataLayout const&) in Instructions.cpp.o
  "llvm::TerminatorInst::getSuccessor(unsigned int) const", referenced from:
      llvm::TerminatorInst::SuccIterator<llvm::TerminatorInst const*, llvm::BasicBlock const>::operator*() const in Structure.cpp.o
  "llvm::TerminatorInst::getNumSuccessors() const", referenced from:
      llvm::TerminatorInst::SuccIterator<llvm::TerminatorInst const*, llvm::BasicBlock const>::SuccIterator(llvm::TerminatorInst const*, bool) in Structure.cpp.o
  "llvm::DomTreeNodeBase<llvm::BasicBlock>::getBlock() const", referenced from:
      dg::analysis::pta::getBasicBlocksInDominatorOrder(llvm::Function&) in PointerSubgraph.cpp.o
  "llvm::GetElementPtrInst::accumulateConstantOffset(llvm::DataLayout const&, llvm::APInt&) const", referenced from:
      dg::analysis::pta::LLVMPointerSubgraphBuilder::handleConstantGep(llvm::GetElementPtrInst const*) in Constants.cpp.o
      dg::analysis::pta::LLVMPointerSubgraphBuilder::createGEP(llvm::Instruction const*) in Instructions.cpp.o
  "llvm::DominatorTreeWrapperPass::verifyAnalysis() const", referenced from:
      dg::analysis::pta::getBasicBlocksInDominatorOrder(llvm::Function&) in PointerSubgraph.cpp.o
  "llvm::Use::getUser() const", referenced from:
      dg::analysis::pta::LLVMPointerSubgraphBuilder::addArgumentOperands(llvm::Function const*, dg::analysis::pta::PSNode*, int) in PointerSubgraph.cpp.o
      dg::analysis::pta::LLVMPointerSubgraphBuilder::addVariadicArgumentOperands(llvm::Function const*, dg::analysis::pta::PSNode*) in PointerSubgraph.cpp.o
      dg::analysis::pta::LLVMPointerSubgraphBuilder::addReturnNodeOperand(llvm::Function const*, dg::analysis::pta::PSNode*) in PointerSubgraph.cpp.o
  "llvm::Type::isSizedDerivedType(llvm::SmallPtrSetImpl<llvm::Type*>*) const", referenced from:
      llvm::Type::isSized(llvm::SmallPtrSetImpl<llvm::Type*>*) const in PointerSubgraph.cpp.o
      llvm::Type::isSized(llvm::SmallPtrSetImpl<llvm::Type*>*) const in Globals.cpp.o
      llvm::Type::isSized(llvm::SmallPtrSetImpl<llvm::Type*>*) const in Instructions.cpp.o
  "llvm::Type::canLosslesslyBitCastTo(llvm::Type*) const", referenced from:
      dg::llvmutils::callIsCompatible(llvm::Function const*, llvm::CallInst const*) in PointerSubgraph.cpp.o
  "llvm::APInt::countLeadingZerosSlowCase() const", referenced from:
      llvm::APInt::operator!() const in PointerSubgraph.cpp.o
      llvm::APInt::countLeadingZeros() const in Constants.cpp.o
      llvm::APInt::countLeadingZeros() const in Instructions.cpp.o
      llvm::APInt::operator!() const in Instructions.cpp.o
      llvm::APInt::countLeadingZeros() const in Calls.cpp.o
      llvm::APInt::operator!() const in Calls.cpp.o
  "llvm::Value::stripPointerCasts() const", referenced from:
      dg::analysis::pta::isRelevantCall(llvm::Instruction const*, bool) in PointerSubgraph.cpp.o
      dg::analysis::pta::LLVMPointerSubgraphBuilder::handleConstantBitCast(llvm::BitCastInst const*) in Constants.cpp.o
      llvm::Value::stripPointerCasts() in Instructions.cpp.o
      dg::analysis::pta::LLVMPointerSubgraphBuilder::createCall(llvm::Instruction const*) in Calls.cpp.o
  "llvm::Value::stripInBoundsOffsets() const", referenced from:
      llvm::Value::stripInBoundsOffsets() in PointerSubgraph.cpp.o
      llvm::Value::stripInBoundsOffsets() in Calls.cpp.o
  "llvm::Value::assertModuleIsMaterializedImpl() const", referenced from:
      llvm::Value::assertModuleIsMaterialized() const in PointerSubgraph.cpp.o
  "llvm::Value::print(llvm::raw_ostream&, bool) const", referenced from:
      llvm::operator<<(llvm::raw_ostream&, llvm::Value const&) in PointerSubgraph.cpp.o
      llvm::operator<<(llvm::raw_ostream&, llvm::Value const&) in Globals.cpp.o
      llvm::operator<<(llvm::raw_ostream&, llvm::Value const&) in Constants.cpp.o
      llvm::operator<<(llvm::raw_ostream&, llvm::Value const&) in Instructions.cpp.o
      llvm::operator<<(llvm::raw_ostream&, llvm::Value const&) in Calls.cpp.o
  "llvm::Value::getName() const", referenced from:
      dg::analysis::pta::LLVMPointerSubgraphBuilder::buildFunction(llvm::Function const&) in PointerSubgraph.cpp.o
      dg::analysis::pta::isRelevantCall(llvm::Instruction const*, bool) in PointerSubgraph.cpp.o
      dg::getMemAllocationFunc(llvm::Function const*) in PointerSubgraph.cpp.o
      dg::analysis::pta::LLVMPointerSubgraphBuilder::createFunctionCall(llvm::CallInst const*, llvm::Function const*) in Calls.cpp.o
      dg::getMemAllocationFunc(llvm::Function const*) in Calls.cpp.o
  "llvm::Module::getFunction(llvm::StringRef) const", referenced from:
      dg::analysis::pta::LLVMPointerSubgraphBuilder::buildLLVMPointerSubgraph() in PointerSubgraph.cpp.o
  "llvm::CastInst::isLosslessCast() const", referenced from:
      dg::analysis::pta::LLVMPointerSubgraphBuilder::handleConstantBitCast(llvm::BitCastInst const*) in Constants.cpp.o
  "llvm::Constant::isNullValue() const", referenced from:
      dg::analysis::pta::LLVMPointerSubgraphBuilder::handleGlobalVariableInitializer(llvm::Constant const*, dg::analysis::pta::PSNodeAlloc*, dg::analysis::pta::PSNode*, unsigned long long) in Globals.cpp.o
  "llvm::Function::BuildLazyArguments() const", referenced from:
      llvm::Function::CheckLazyArguments() const in PointerSubgraph.cpp.o
      llvm::Function::CheckLazyArguments() const in Structure.cpp.o
  "vtable for llvm::FunctionPass", referenced from:
      llvm::FunctionPass::FunctionPass(char&) in PointerSubgraph.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "vtable for llvm::DominatorTreeWrapperPass", referenced from:
      llvm::DominatorTreeWrapperPass::DominatorTreeWrapperPass() in PointerSubgraph.cpp.o
      llvm::DominatorTreeWrapperPass::~DominatorTreeWrapperPass() in PointerSubgraph.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
  "vtable for llvm::Pass", referenced from:
      llvm::Pass::Pass(llvm::PassKind, char&) in PointerSubgraph.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [lib/libLLVMpta.dylib] Error 1
make[1]: *** [lib/CMakeFiles/LLVMpta.dir/all] Error 2
make: *** [all] Error 2
```

I do generate the Makefile with the following command:

```
LLVM_DIR=$HOME/Documents/llvm61/build/lib/cmake/llvm cmake -H. -Bbuild
```

Is there anything that I can do to compile **dg**?

EDIT: Is OSX supported?

Regards

--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -98,9 +98,36 @@ add_library(LLVMdg SHARED
 	llvm/analysis/DefUse/DefUse.cpp
 )
 
-target_link_libraries(LLVMdg
-			PUBLIC LLVMpta
-			PUBLIC LLVMrd)
+# Get proper shared-library behavior (where symbols are not necessarily
+# resolved when the shared library is linked) on OS X.
+if(APPLE)
+  set_target_properties(LLVMdg PROPERTIES
+    LINK_FLAGS "-undefined dynamic_lookup"
+  )
+
+  set_target_properties(LLVMpta PROPERTIES
+    LINK_FLAGS "-undefined dynamic_lookup"
+  )
+
+  set_target_properties(LLVMrd PROPERTIES
+    LINK_FLAGS "-undefined dynamic_lookup"
+  )
+endif(APPLE)
+
+if (APPLE)
+	target_link_libraries(LLVMdg
+				PUBLIC LLVMpta
+				PUBLIC LLVMrd
+				PRIVATE ${llvm_support}
+				PRIVATE ${llvm_analysis}
+				PRIVATE ${llvm_irreader}
+				PRIVATE ${llvm_bitwriter}
+				PRIVATE ${llvm_core})
+else()
+	target_link_libraries(LLVMdg
+				PUBLIC LLVMpta
+				PUBLIC LLVMrd)
+endif(APPLE)
 
 install(TARGETS LLVMdg LLVMpta LLVMrd PTA RD DGAnalysis
 	LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR})

